crosscluster/logical: run offline init scan if creating new tables

msbutler opened the pull request.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/137401)
<!-- Reviewable:end -->

msbutler committed:
crosscluster/physical: refactor pcr ingestion for ldr init scan use

Epic: none

Release note: none

msbutler committed:
crosscluster/logical: run offline init scan if creating new tables

This patch enables LDR to run an offline initial scan via the pcr ingestion
path on newly created tables. In the future, we should consider pre splitting
and scattering the key space as well.

Epic: none

Release note: none

msbutler commented on pkg/crosscluster/logical/logical_replication_job.go:
```diff
@@ -582,6 +715,12 @@ func (rh *rowHandler) handleRow(ctx context.Context, row tree.Datums) error {
+		// NB: the frontier update for the replicated time may not get sent over to
+		// the source, but this isn't a big deal-- when the distsql flow retries a
+		// new heartbeat will be sent.
+		return errOfflineInitialScanComplete
```
Hrm. Actually, i think we do need to bind the origin timestamp on the replicated data.

msbutler committed:
kv/bulk: add AddMVCCKeyLDR to sst_batcher

Epic: none

Release note: none

msbutler commented on pkg/kv/bulk/sst_batcher.go:
```diff
@@ -345,6 +345,25 @@ func (b *SSTBatcher) AddMVCCKeyWithImportEpoch(
 	return b.AddMVCCKey(ctx, key, b.valueScratch)
 }
 
+func (b *SSTBatcher) AddMVCCKeyLDR(
```
i can add a little unit test in sst_batchet_test.go, if folks like this approach.

msbutler committed:
crosscluster/producer: add initial scan only event stream

msbutler committed:
ammend to sip changes

msbutler committed:
ldr job

msbutler committed:
errors is exp

msbutler committed:
debug

msbutler committed:
defer heartbeat sender in exec plan
